scipy/optimize/tests/test_optimize.py

```diff
@@ -463,6 +511,87 @@ def test_neldermead_adaptive():
                     options={'adaptive':True})
     assert_equal(res.success, True)
 
+def test_bounded_powell_outsidebounds():
+    # One cool feature of the bounded Powell method is that if you
+    # start outside the bounds, the final answer will still be
+    # within the bounds (provided that the user doesn't make a bad
+    # choice for the `direc` argument). 
+    func = lambda x: np.sum(x**2)
+    bounds = (-1, 1), (-1, 1), (-1, 1)
+    x0 = [-4, .5, -.8]
+
+    # we're starting outside the bounds, so we should get a warning
+    with assert_warns(optimize.OptimizeWarning):
+        res = optimize.minimize(func, x0, bounds=bounds, method="Powell")
+    assert_allclose(res.x, np.array([0.] * len(x0)), atol=1e-6)
+    assert_equal(res.success, True)
+    assert_equal(res.status, 0)
+
+    # However, now if we change the `direc` argument such that the
+    # set of vectors does not span the parameter space, then we may
+    # not end up back within the bounds. Here we see that the first
+    # parameter cannot be updated!
+    direc = [[0, 0, 0], [0, 1, 0], [0, 0, 1]]
+    # we're starting outside the bounds, so we should get a warning
+    with assert_warns(optimize.OptimizeWarning):
+        res = optimize.minimize(func, x0, 
+                                bounds=bounds, method="Powell",
+                                options=dict(direc=direc))
+    assert_allclose(res.x, np.array([-4., 0, 0]), atol=1e-6)
+    assert_equal(res.success, False)
+    assert_equal(res.status, 3)
+
+def test_bounded_powell_vs_powell():
+    # here we test an example where the bounded Powell method
+    # will return a different result than the standard Powell
+    # method.
+
+    # first we test a simple example where the minimum is at
+    # the origin and the minimum that is within the bounds is
+    # larger than the minimum at the origin.
+    func = lambda x: np.sum(x**2)
+    bounds = (-5, -1), (-10, -.1), (1, 9.2), (-4, 7.6), (-15.9, -2)
```

Comment:
there's no tests with a mixture of non-finite bounds. i.e. handling mixtures of these:
```
(-np.inf, -1)
(-np.inf, np.inf)
(1, np.inf)
(1., 9.2)
```